prov/rxd/src/rxd_av.c

```diff
@@ -185,25 +142,48 @@ static int rxd_av_insert(struct fid_av *av_fid, const void *addr, size_t count,
 	}
 
 	for (; i < count; i++, addr = (uint8_t *) addr + av->dg_addrlen) {
-		node = ofi_rbmap_find(&av->rbmap, (void *) addr);
-		if (node) {
-			rxd_addr = (fi_addr_t) node->data;
-		} else {
-			ret = rxd_av_insert_dg_addr(av, addr, &rxd_addr,
-						    flags, context);
-			if (ret)
-				break;
-		}
-
-		util_addr = av->rxd_addr_table[rxd_addr].fi_addr == FI_ADDR_UNSPEC ?
-			    rxd_set_fi_addr(av, rxd_addr) :
-			    av->rxd_addr_table[rxd_addr].fi_addr;
+		
+        ret = rxd_av_insert_dg_addr(av, addr, &dg_addr, flags, context);
+        if (ret)
+            break;
+        ret = rxd_av_insert_fi_addr(av, &dg_addr, &util_addr , flags);
+        if (ret) {
+            fi_av_remove(av->dg_av, &dg_addr, 1, flags);
+            break;    
+        }
+        dlist_foreach(&av->util_av.ep_list, av_entry) {
+
+            util_ep = container_of(av_entry, struct util_ep, av_entry);
+            rxd_ep =  container_of(util_ep, struct rxd_ep, util_ep);
+
+            // function to get peer_entry from ep->rxd_peer_pool
+            peer_entry = rxd_get_peer_entry(rxd_ep,(void*)addr);
+
+            if (peer_entry->dg_addr != FI_ADDR_UNSPEC){
+                fi_av_remove(av->dg_av, &dg_addr, 1, flags);
+                ofi_av_remove_addr(&av->util_av, *fi_addr);
+                goto fail;
+            }
+            
+            peer_entry->fi_addr =  util_addr;
+            peer_entry->dg_addr = dg_addr;
+            peer_entry->peer_addr = ofi_buf_index(peer_entry);
+        
+            // for each fi_addr create a unique fi_addr-->rxd_addr per ep.
+            ret = rxd_fiaddr_hash_insert(rxd_ep, fi_addr, &(peer_entry->peer_addr));                       
+            if (ret){
+                fi_av_remove(av->dg_av, &dg_addr, 1, flags);
+                ofi_av_remove_addr(&av->util_av, *fi_addr);
+                goto fail;
+            }          
+        }               
```

Comment:
I'm not really sure what's happening here. It feels like you need a function rxd_init_ep_peer or something that does the initialization if it's not there and if it is there, updates values or something. This is confusing

Reply:
```suggestion

            if (peer_entry->dg_addr != FI_ADDR_UNSPEC){
                fi_av_remove(av->dg_av, &dg_addr, 1, flags);
                ofi_av_remove_addr(&av->util_av, *fi_addr);
                goto fail;
            }
            
            peer_entry->fi_addr =  util_addr;
            peer_entry->dg_addr = dg_addr;
            peer_entry->peer_addr = ofi_buf_index(peer_entry);
        
            // for each fi_addr create a unique fi_addr-->rxd_addr per ep.
            ret = rxd_fiaddr_hash_insert(rxd_ep, fi_addr, &(peer_entry->peer_addr));                       
            if (ret){
                fi_av_remove(av->dg_av, &dg_addr, 1, flags);
                ofi_av_remove_addr(&av->util_av, *fi_addr);
                goto fail;
            }          
        }               
```
```suggestion

            if (peer_entry->dg_addr != FI_ADDR_UNSPEC){
                fi_av_remove(av->dg_av, &dg_addr, 1, flags);
                ofi_av_remove_addr(&av->util_av, *fi_addr);
                goto fail;
            }
            
            peer_entry->fi_addr =  util_addr;
            peer_entry->dg_addr = dg_addr;
            peer_entry->peer_addr = ofi_buf_index(peer_entry);
        
            // for each fi_addr create a unique fi_addr-->rxd_addr per ep.
            ret = rxd_fiaddr_hash_insert(rxd_ep, fi_addr, &(peer_entry->peer_addr));                       
            if (ret){
                fi_av_remove(av->dg_av, &dg_addr, 1, flags);
                ofi_av_remove_addr(&av->util_av, *fi_addr);
                goto fail;
            }          
        }               
```